text/0680-implicit-injection-deprecation.md

```diff
@@ -0,0 +1,127 @@
+   For example, `ember-data` has laid out its plan for adopting only specific packages in [Project Trim](https://github.com/emberjs/data/issues/6166).
+   However, because `store` is injected by default on all Ember.Route and Ember.Controller factory objects,
+   a user cannot easily opt out of `@ember-data/store`. This is a common problem space in other communities as well.
+   With previous JetBrains IDE installations, adding plugins required a reload. However, by removing a contructor function
+   that injected all the dependencies at once, they were able to avoid reload after installation of a plugin.
+
+2. Eager initialization of the dependency tree from which the property came from prevents tree shaking and incurs a
+   performance hit. For example, if you app does not need `ember-data` to render the entry route, the `store` injection will
+   take up more CPU cycles than necessary, hurting common user facing metrics.
+
+3. Eager initialization makes it hard to users to write tests that need to take advantage of a stub.
+   This is often a silent and annoying hurdle that users have to get around when writing tests.
+   Moreover, this is especially apparent for users moving to Ember's new testing APIs in [RFC 232](https://github.com/emberjs/rfcs/blob/master/text/0232-simplify-qunit-testing-api.md) and [RFC 268](https://github.com/emberjs/rfcs/blob/master/text/0268-acceptance-testing-refactor.md),
+   ensuring instance initializers run before each test.
+
+4. Injection on commonly used framework objects can lead to deviation of project specific styleguides. For instance,
+   injecting the `router` on the Ember.Component factory object will certainly lead to instances where the `router` is explicitly
+   injected and other instances where is it implicitly injected.
+
+5. Users have come to associate `Route.model()` with a hook that returns `ember-data` models in the absense of explicit injection.
+   While this can be true, it is not wholly true. New patterns of data loading are becoming
+   accepted in the community including opting to fetch data in a Component.
+
+
+## Detailed design
+
+Removing implicit injections is not possible until we first issue deprecations.
+This helps us maintain our strict semver commitment to not completing a major version bump and
+removing APIs without a deprecation path. As a result, it will take us 3 phases to remove `owner.inject`.
+
+### 1. Deprecate implicit injection on target object
+
+The first step will be issuing a deprecation on the target factory object that is receiving a property. This will surface the deprecation
+on user's owned objects including transitively through addons like `ember-data`. This can be accomplished by installing a native getter
+and setter on the target. Whenever the property is accessed, we will check if the property is explicitly injected. If it isn't,
+we will issue a deprecation with an until value, id and url to read more about this deprecation.
```

Comment:
It would probably be good to mention that these getters/setters will only be setup in `DEBUG` builds, and in production builds we will assign the property just like we do now. The goal is to avoid any production penalty for supporting this deprecation.